File: backend/app/agents/case_discovery.py
import re

from app.clients.courtlistener import CourtListenerClient


def _slugify(name: str) -> str:
	return re.sub(r"[^a-z0-9]+", "-", name.lower()).strip("-")


def _strip_html(text: str) -> str:
	return re.sub(r"<[^>]+>", "", text).strip()


def run(query: str, limit: int = 9) -> list[dict]:
	client = CourtListenerClient()
	try:
		results = client.search_opinions(query, court="scotus", limit=limit)
		cases = []
		for r in results:
			name = r.get("caseName", "").strip()
			if not name:
				continue
			date_filed = r.get("dateFiled", "")
			year = 0
			if date_filed:
				try:
					year = int(date_filed[:4])
				except ValueError:
					pass
			citations = r.get("citation", [])
			snippet = _strip_html(r.get("snippet", "").strip())
			url = r.get("absolute_url", "")
			cases.append({
				"id": _slugify(name),
				"name": name,
				"year": year,
				"category": "SCOTUS",
				"summary": snippet or "No summary available.",
				"citation": citations[0] if citations else "",
				"url": f"https://www.courtlistener.com{url}" if url else "",
				"source": "courtlistener",
			})
		return cases
	finally:
		client.close()

File: frontend/lib/agents/caseDiscovery.ts
import { searchOpinions } from '@/lib/courtlistener';

function slugify(name: string): string
{
	return name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '');
}

function stripHtml(text: string): string
{
	return text.replace(/<[^>]+>/g, '').trim();
}

export async function run(query: string, limit: number = 9): Promise<any[]>
{
	const results = await searchOpinions(query, { court: 'scotus', limit });
	const cases: any[] = [];

	for(const r of results)
	{
		const name = (r.caseName ?? '').trim();
		if(!name) continue;

		const dateFiled: string = r.dateFiled ?? '';
		let year = 0;
		if(dateFiled)
		{
			const parsed = parseInt(dateFiled.slice(0, 4), 10);
			if(!isNaN(parsed)) year = parsed;
		}

		const citations: string[] = r.citation ?? [];
		const snippet = stripHtml((r.snippet ?? '').trim());
		const url: string = r.absolute_url ?? '';

		cases.push({
			id: slugify(name),
			name,
			year,
			category: 'SCOTUS',
			summary: snippet || 'No summary available.',
			citation: citations[0] ?? '',
			url: url ? `https://www.courtlistener.com${url}` : '',
			source: 'courtlistener',
		});
	}

	return cases;
}
